### ISSUE TITLE
SubClass class property assignment lose BaseClass property type

### ISSUE THREAD

**Issue by jeremychone** [Microsoft/TypeScript#18531]
(I did search for this bug, but could not find it. Sorry if I missed anything). 

**TypeScript Version:**  2.5.2

**Code**

```ts
class ClassA {
  val: number = 2;
}

// Works as expected
class ClassB extends ClassA {
  constructor() {
    super();
    this.val = this.val + 1; // this.val type check works as expected
  }
}

// Error as expected
class ClassC extends ClassA {
  val = "hello"; // <-- Error as expected, can't set string as ClassA.val is number
}

// Wrong ERROR (IMO) Should be equivalent to ClassB
class ClassD extends ClassA {
  val = this.val + 1; // <-- ERROR: 'val' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
}
```

**Expected behavior:**

`ClassD` should be the equivalent as `ClassB`. 

```ts
// this.val should be of type "number" as defined in base class (as recognized in ClassB).
class ClassD extends ClassA {
  val = this.val + 1; // <-- should be the exact equivalent as `ClassB`, this.val should be of the typeof BaseClass.val
}
```

**Actual behavior:**
```ts
// Wrong ERROR (IMO) Somehow assignment of "this.val" make "val" of type any now (but worked fine in ClassB)
class ClassD extends ClassA {
  val = this.val + 1; // <-- Wrong ERROR: 'val' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
}
```

#### Practical usecase

While simple re-assigning might seems uncommon, base property augmentation can be common, and the direct class property make the class more tierce. 

Our basic example is as follow:

```ts
// base.ts

export class BaseView {
  events: EventBindings = {};
  config: Config = {};
}

// mono type assign
export function assign<T extends object>(a: T, b: T): T {
  return Object.assign(a, b);
}
```

```ts
// SubView.ts
import {assign, BaseView} from 'base';

export class SubView extends { 

  events = assign(this.events, { // <-- Unfortunately, type error here, "this.events" has become "any"
    "click": () => { /* do something */ }
  }  
  config = assign(this.config, { // <-- Unfortunately, type error here, "this.config" has become "any"
    modal: true
  }  
}
```

Right now, we put those "augmentations" in constructosr, but would be nicer if we can just have them as direct property assignment.

**Comment by mhegazy** [comment]
Please see https://github.com/Microsoft/TypeScript/issues/3667

**Comment by jeremychone** [comment]
@mhegazy sorry. Too bad it cannot be fixed.

**Comment by mhegazy** [comment]
I would not give up hope just yet. we have not found a solution that is not a breaking change, but we are still looking into it.

**Comment by jeremychone** [comment]
@mhegazy Cool, thanks for the hopeful note.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
